### ISSUE TITLE
Replace Database layer with doctrine/dbal

### ISSUE THREAD

**Issue by dhensby** [silverstripe/silverstripe-framework#6632]
More and more we're seeing issues raised about our DB layer and it not working under specific circumstances. Our DB layer is also in need of a slight refactor to properly separate concerns and stop us hard coding logic related to DB drivers into our generic connectors.

Rather than spending our time re-factoring our DB connectors and maintaining an incomplete set of DB drivers I suggest we move our DB layer to a thirdparty - this will have the advantages of:

1. Reducing maintenance burden on core team - We currently struggle with the maintenance of external DB adapters especially MSSQL
2. Increase support of other DB drivers - The number of DB drivers we offer is limited and can be increased by moving to a thirdparty lib

I've looked at [Laravel's Illuminate](https://github.com/illuminate/database) and [Doctrine's DBAL](https://github.com/doctrine/dbal). Laravel's database component is far too tightly coupled to the Eloquent ORM - we'd have to pull it in even though we aren't going to use it.

Looking into doctrine/dbal and it's got a really nice API and a very similar architecture to our current DB layer. It looks like a pretty awesome and well architected DB layer as well as giving us the advantage of a lot of stock drivers and any other open-source drivers that have subsequently been written.

If we're able to swap out our DB layer, we'll drop a lot of hand-rolled code whilst improving functionality and focusing on adding value.

Other advantages:

1. More easily move to database migrations (https://github.com/doctrine/migrations/)
2. Defer DB connections to be "on demand" rather than connecting on every request

Issues:

1. There is no Enum support in doctrine/dbal and so we'll have to move this logic to the application layer.\*

\*Personally, I think this is something that we should be doing anyway as Enums can be painful for a number of reasons see [doctrine's docs on enums](http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/cookbook/mysql-enums.html)

**Comment by sminnee** [comment]
Dan, could you outline which classes you'd see as being replaced by dbal, and where the coupling points to dbal woud be?

**Comment by dhensby** [comment]
I think this work is something that will need to be done in steps. In the long term I'd like to see pretty much everything in our `SilverStripe\ORM\Connect` namespace go. However, this is a lot of work and I'm starting by aiming at keeping most of our SQL generation and simply using doctrine/dbal to execute the queries for us.

The more I'm looking into doctirne/dbal the more I'm seeing how powerful it is and that it appears that it will solve some problems for us or, at least, make it simpler to do things we do now. Automatica aliasing of table names is one example.

I don't see a big change in coupling points; anywhere we currently couple to `DBQueryBuilder` we'd likely replace with the dbal query builder. So the main coupling points would be `DataQuery`, the `DBField` classes (they need to present represent themselves as columns in the DB table, as they do already) and `SilverStripe\ORM\Filters` would also need to be able to build `WHERE` clauses.

**Comment by sminnee** [comment]
Ok I'm not opposed to this, but I'm not sure I'd rate its priority highly for SS4. My instinct is that it's probably more work to do this that fix the issues we already know about.

That said, if you were going to drive this, you'd get to make the priority call. ;-)

I've got mixed feelings about doing a half-job of this to hit the SS4 timeline too. It depends a bit on what a "half job" would look like but I'd be more keen to leave SS4 in a consistent state than to just try and squeeze something in.

I'd also note that I'm much more nervous about replacing the ORM, but you seem to be more or less in agreement with that too.

**Comment by dhensby** [comment]
> My instinct is that it's probably more work to do this that fix the issues we already know about.

That may well be the case; but I think this is the wrong way to approach. Whilst the cost to solve the known issues now may be less than integrating a 3rd party lib, the integration will pay-off in the long run, especially if SS4 is going to be an LTS release.

Having said that, I don't think it should be a priority for SS4 - there's only so much that can be done in time for the release.

I've been taking a look at integrating it in my evenings and it does look very nice and I've got the CMS loading and so on, so I'll spend some more time looking into it but with no commitment to being able to complete for the beta.

**Comment by dhensby** [comment]
RE the ORM - yes, I don't think that's anything we can remove any time soon.

Laravel's is the closest in terms of syntax and familiarity, but there's no MTI. No matter what we'd be maintaining our own extension of another ORM if that were ever to happen... I'm focused on the areas where we have little differentiation from other libraries.

Right now, doctrine/dbal looks like a great fit.

**Comment by stevie-mayhew** [comment]
I haven't used either of those database connectors extensively so can't really comment on the specifics for each, but I think that the idea of changing out the DB connector is a good one for the exact reasons you've outlined.

**Comment by sminnee** [comment]
I'd merge a PR if you did this, but it won't get into open sourcerers' backlog.

**Comment by sminnee** [comment]
This won't block the beta1 milestone. If you manage to do this, then great, otherwise it'll likely be an SS5 item.

**Comment by dhensby** [comment]
From my initial investigation work on this there are a few key areas where doctrine/dbal doesn't work quite the same way that our DB layer does or has "missing features".

Potential blockers:
1. [Renaming columns doesn't seem to be possible in the DBAL](https://github.com/doctrine/dbal/blob/v2.5.12/lib/Doctrine/DBAL/Schema/Table.php#L338-L343)
 2. Fixing table case is not currently possible because change detection on tables is done in a case-insensitive manner
3. No abstract way to get the `RAND()` method for a DBMS
4. Can't find a cross DBMS way to select a database. When connecting to a DB, if we supply the DB name and it doesn't exist, we get an error - so we have to reconnect without the DB name and then create and select it, this appears problematic at the moment.
5. Enums are not supported
6. Can't set a `select` statement to be `DISTINCT` in the query builder

High effort upgrades:
1. Move DB query building to the DBAL query builder (though creating our own DQL builder may be better)
2. Have to use `Convert::symbol2sql()` for quoting columns (ANSI mode is not on by default with dbal because of "inconsistencies" so we can't hard code `"` as our identifier quote character)
3. Can't "convert" an `SQLExpression` object from one type to another (currently we can build an `SQLUpdate` query and turn it into an `SQLSelect`)

Benefits of doctrine/dbal:
1. Closely aligned APIs for DB abstraction (similar architecture of classes as well as method names)
2. On demand DB connection (doesn't connect until a query needs to be run)
3. Good typing of data pulled out of the DB (currently all data is pulled out as string, dbal will type it correctly for us)
4. Simple to use query builder with easy parametrisation including support for: array expansion, and named and positional parameter tracking.

**Comment by sminnee** [comment]
OK, based on that feedback, particularly point 2 of "high-effort upgrades", I can't see this happening in SS4.

On the potential blockers:

 * 1: I'm not sure that we use renaming in dev/build, and if we do, I expect it's only to shuffle obsolete columns out the way. If people use it in their code it'll just be for `requireDefaultRecords()` migrations, and presumably we could switch to a migration system based on dbal
 * 2: If that means that tables are consistently accessible in a case-insensitive manner then perhaps that's okay?
 * 3: Seems like a PR we could supply to DBAL, although I should note that `ORDER BY RAND()`, which is probably our use case, is an anti-pattern (prevents use of indexes) and it might be better to avoid its use.
 *  4: I don't see an issue with `try { /* connect */ } catch { /* create */ /* connect */ }`. The code around switching databases for temp dbs and the like can do with a clean-up
 * 5: Guh, WTF? PR to DBAL seems like the way ahead there.
 * 6: We can probably do better than these `DISTINCT` queries, with tighter GROUP BY controls. We may already have that in place (the DISTINCT goes way back to SS2.0 days from memory).

**Comment by sminnee** [comment]
On the benefits: I see point 2 as being achieved by moving the database to an Injector service, which is something that is much more likely for SS4.

**Comment by sminnee** [comment]
If the ORM were split into a separate composer package, then you could potentially use `dev-master` of that package and the 4.x versions of all other packages, which would mean that a re-write based on this would be something you'd have to wait as long to use.

**Comment by dhensby** [comment]
Blockers:

1. We use it to name columns `_obsolete_[colname]`. I'm not really sure why this isn't in the dbal core - I understand that from schema migration you can't tell if a col is renamed or just removed, but you should still be able to rename a column explicity IMO. There's nothing stopping us implementing a column renaming method ourselves.
2. I'm afraid it doesn't. We'd have to look at some other kind of work around.
3. Yes, they probably don't have it for this reason, it's not great practice, we should probably discourage its usage
4. Agreed.
5. [Doctrine have explained their logic](http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/cookbook/mysql-enums.html) and I agree with their points. I've thought for some time we should move the enum DBField to the application layer - this isn't really a blocker, we just need to implement it in the app layer
6. agreed

> If the ORM were split into a separate composer package, then you could potentially use dev-master of that package and the 4.x versions of all other packages, which would mean that a re-write based on this would be something you'd have to wait as long to use.

That's true and I'd like that myself, but we'd have to decide how granular we really want to split the framework... A discussion for another thread, perhaps.

**Comment by dhensby** [comment]
In terms of high effort upgrade points, really the solution is to improve the abstraction layer (`DataQuery`) between our ORM and the DB queries so that no one is writing SQL in the application. This will remove the need for anyone to be touching `Convert::symbol2sql()`, devs won't need the DB QueryBuilder nor the `SQLConditionalExpression` class.

**Comment by wakes** [comment]
Have you looked at readbeanphp http://redbeanphp.com/index.php? Very lightweight, manages the schema for you (unless you say 'freeze' e.g. in test/live environments), I've used in play code and great for prototyping, not sure re production, performance etc.

**Comment by Firesphere** [comment]
I've had a few plays with redbean, it's pretty nice, but I'm not well-informed enough to know if it suits SilverStripe

**Comment by dhensby** [comment]
I haven't looked at redbean.

I've take a quick browse now but it doesn't look quite right for SS. doctrine/dbal works really well with our current approach

**Comment by tractorcow** [comment]
Summary of stuff @dhensby and I discussed on slack:

 - No more quotes for identifiers anywhere; ANSI sql or not
 - Shift all query objects to properly constructed object representation of SQL queries that can be augmented in a type-gnostic fashion; No more SQL strings anymore.
 - field / table rewriting must still work (e.g. for versioned / fluent) but it will operate on specific logical steps instead of string manipulations. E.g. rewrite field vs rewrite string
 - Middleware for object -> sql generation
 - All identifiers lowercase, some field name (foreignkey_id) conventions changed.
 - DB abstraction will probably be shifted to inside doctrine/dbal instead of outside it.
 - I think it would be cool to do closure -> sql transpiling, probably never going to get around to it. :)
 
 On enums, they aren't supported well in many DB connectors anyway. I don't mind it being supported above the DB level if necessary.

**Comment by tractorcow** [comment]
Apparently quoting happens internally in doctrine so not anything we need to worry about. :)

I'm not sure I understand the significance of capital letters in identifiers. @dhensby says we need to use all lowercase from here on anyway.

**Comment by dhensby** [comment]
> As such I'd recommend that anything that auto-generates SQL only adds quotes to identifiers that contain a capital letter or other reserved character, rather than avoiding them entirely.

doctrine/dbal does this automatically (at least for reserved words).

You can see their response to a ticket I opened a while ago here: https://github.com/doctrine/dbal/issues/2710#issuecomment-299854401

>for example in the code that migrates from the old identifier names to the new.

We will need to write this kind of migration manually, I'm afraid. The dbal layer doesn't provide a way to distinguish between different cased identifiers in the abstraction layer. This could be part of upgrader 2.0.

It'll also spell the end of case correction logic for table names / columns.

**Comment by robbieaverill** [comment]
I've raised an RFC to remove support for enums at #8401

**Comment by sminnee** [comment]
This hasn't been touched in a few years, and I want to flag that I question whether the benefit of this will _ever_ be worth the upgrade pain that it would inevitably cause. The fact is that the current base access layer, while not perfect, is not in my view a particularly large source of either bugs or or maintenance (feel free to challenge that).

I noted on #9520 a comment from @dnsl48:
> I agree with @dhensby that it may not be worthwhile without introducing a proper dbal first.

As such, I wonder if keeping this ticket open as a "someday, maybe" item is preventing appropriate decision-making on other issues.

**Comment by dnsl48** [comment]
Regarding #9520, I don't think this issue affected my judgement :)
I was rather thinking about the suggested feature increasing complexity of the existing code base and its maintainability.

**Comment by sminnee** [comment]
Ah, right :)

**Comment by dhensby** [comment]
I agree there probably isn't much value in keeping this open. It's a huge amount of work. I've actually had a DBAL rewrite done and then got most of the tests green only to find our approach of using case sensitive column names was actually a bit of a problem and that using ansi-quotes was also an issue. This drastically increased the scope of the work and I haven't been able to progress it since then.

To get a new DB layer in we actually need a proper domain specific querying layer (which we almost have with DataQuery) except that we also allow raw SQL to be injected and that needs to be replaced with an abstracted layer.

All-in-all, it's going to need some proper concerted effort and a hell-of-a-lot of cross-module breaking changes...

**Comment by GuySartorelli** [comment]
> I've actually had a DBAL rewrite done

@dhensby Did you happen to have that lying around somewhere still?

**Comment by dhensby** [comment]
@GuySartorelli [yep](https://github.com/dhensby/silverstripe-framework/tree/pulls/4/dbal). It's very old, though 😅 

The problem I came up against was the dbal doesn't actually support case-sensitive database identifiers (column names, etc), which is pretty fundamental to the conventions of SS schemas. The problem is that for the widest possible database support, these identifiers should be lowercased. There were a couple of issues that I opened in the dbal library which were closed:

- [Schema checks are case-insensitive](https://github.com/doctrine/dbal/issues/2710) - I think that was needed for our feature where we re-case column names on dev/build if we see they are bad
- [Auto increment column names not case sensitive](https://github.com/doctrine/dbal/issues/2714) - I did get this fixed, but it was just a symptom of a wider problem and set of assumptions within the library.

I think I did get it passing with mysql and the default framework, but once I tried to get postgres working it became clear I just wouldn't be able to make this work without a fundamental change to the SS conventions and have all identifiers be lowercased. That could be done nicely with a mapping layer between ORM and queries (which already existed and could be used), but it would mean anyone using manual queries is going to have a bad time. Ultimately, I spent enough time on it that I decided perusing it any further wasn't worth it because the benefits were unlikely to outweigh the costs.

Ultimately I was hoping we could use the dbal library as a drop-in replacement for our own DB layer, which would unlock things like migrations and wider DBMS support, but it looked more and more like a fundamental shift in the DB integration was going to be needed instead.

**Comment by GuySartorelli** [comment]
Thank you, that's very useful. I've been looking at the doctrine dbal recently more for my own learning than anything, and this issue kept popping up in my head. If we _could_ make this shift it could (depending on the implementation) result in a lot of code we no longer need to maintain (along with the benefits you've already noted).

Am I being naive thinking that the case sensitivity problem could be resolved by just enforcing table and column names to be lower case, and providing a migration to update the schema of existing databases?

Any custom SQL queries _should_ be using abstractions like `DataObject::getSchema()->sqlColumnForField($class, $field)`, though I recognise that not everyone will be using that so some queries will need to be manually updated to avoid hardcoded table/column names. I assume that's what you meant by:

>  That could be done nicely with a mapping layer between ORM and queries (which already existed and could be used), but it would mean anyone using manual queries is going to have a bad time.

It's unlikely that we would make this change in CMS 7 and probably not even 8 (there are more important fish to fry), but I would like to make this change eventually. If we do decide to make the change I think a combination of migrating the schema, and broadcasting the intention and known painpoints of this early ("if you have hardcoded table or column names in your custom SQL queries, you will need to use the abstractions such as `sqlColumnForField` instead") will go a long way to resolving problems.

> it looked more and more like a fundamental shift in the DB integration was going to be needed instead.

Other than the case sensitivity problem, what fundamental changes would be needed if you remember? You mentioned some potential blockers in https://github.com/silverstripe/silverstripe-framework/issues/6632#issuecomment-285057704 but most of those have solutions discussed in comments in this issue.

**Comment by dhensby** [comment]
I'm really picking at old memories and so I'm bound to forget things, but functionally I think the two blocking issues were ansi quotes and lowercase identifiers.

At the time of my work on the DBAL replacement we were coming out of the SS4 upgrades and the learnings around what such big upgrades looked like for users and the commercial implications. The feeling at the time was simply it was not sustainable to release such huge changes frequently because there was often not the budget for those kinds of upgrades and it would just lead to more fragmentation of versions being used. We got lots of feedback from the community about the pains and cost of SS3->SS4 upgrades. Ultimately the 3rd blocker was political will to make the change and accept the potential disruption, I certainly put a self-enforced limitation that the replacement should be a transparent drop-in replacement.

The suggestions you make around good forward guidance and planning ahead is needed, but I'm out of the loop now on upgrade paths for SS and whether concerns about potentially disruptive upgrades weigh quite so heavily.

Like you, I see a real advantage in replacing this part of the framework we don't need to author ourselves and can source elsewhere. That was one of my big pushes at the time; why maintain our own implementation of all these core pieces of functionality when we can make use of specialised ones being maintained by people far more knowledgeable in that focused domain? I think I made the point at the time that the DBAL library had more contributions than all of SS and it was only concerned with DB abstraction whereas our DB layer had about 3 contributors over all time. It seems like an easy win for maintainability and stability to make the switch!
